cmd/orchestrate/orchestrate.go

```diff
@@ -0,0 +1,314 @@
+			// TODO fix resend mechanism
+			log.Errorln("We need to fix this resend stuff ...")
+		}
+		if err := delivered.Ack(false); err != nil {
+			log.Errorf("failed to ack message for reason: %v", err)
+		}
+	}
+}
+
+// schemaNameFromQueue returns the schema to use for messages
+// determined by the queue
+func schemaNameFromQueue(queue string, body []byte) (string, error) {
+	if queue == queueInbox {
+		return schemaFromInboxOperation(body)
+	}
+	m := map[string]string{
+		queueVerify:    "ingestion-accession-request",
+		queueComplete:  "ingestion-completion",
+		queueIngest:    "ingestion-trigger",
+		queueMapping:   "dataset-mapping",
+		queueAccession: "ingestion-accession",
+		queueBackup:    "ingestion-completion",
+	}
+
+	if m[queue] != "" {
+		return m[queue], nil
+	}
+
+	return "", fmt.Errorf("Don't know what schema to use for %s", queue)
+}
+
+// schemaFromInboxOperation returns the operation done in inbox
+// supplied in body of the message
+func schemaFromInboxOperation(body []byte) (string, error) {
+	message := make(map[string]interface{})
+	err := json.Unmarshal(body, &message)
+	if err != nil {
+		return "", err
+	}
+
+	inboxOperationFetch, ok := message["operation"]
+	if !ok {
+		return "", errors.New("Malformed message, inbox operation is missing")
+	}
+
+	inboxOps, ok := inboxOperationFetch.(string)
+	if !ok {
+		return "", errors.New("Could not cast operation attribute to string")
+	}
+
+	switch inboxOps {
+	case "upload":
+		return "inbox-upload", nil
+	case "rename":
+		return "inbox-rename", nil
+	case "remove":
+		return "inbox-remove", nil
+	default:
+		return "", errors.New("Could not recognize inbox operation")
+	}
+
+}
+
+func ingestMessage(body []byte) ([]byte, interface{}) {
+	var message upload
+	err := json.Unmarshal(body, &message)
+	if err != nil {
+		return nil, nil
+	}
+
+	msg := trigger{
+		Type:               "ingest",
+		User:               message.User,
+		Filepath:           message.Filepath,
+		EncryptedChecksums: message.EncryptedChecksums,
+	}
+
+	publish, _ := json.Marshal(&msg)
+
+	return publish, new(trigger)
+}
+
+func finalizeMessage(body []byte, conf *config.Config) ([]byte, interface{}) {
+	var message request
+	err := json.Unmarshal(body, &message)
+	if err != nil {
+		return nil, nil
+	}
+	accessionID := uuid.NewSHA1(
+		uuid.NewSHA1(uuid.NameSpaceDNS, []byte(conf.Orchestrator.ProjectFQDN)),
+		body).URN()
+
+	msg := finalize{
+		Type:               "accession",
+		User:               message.User,
+		Filepath:           message.Filepath,
+		DecryptedChecksums: message.DecryptedChecksums,
+		AccessionID:        accessionID,
+	}
+
+	publish, _ := json.Marshal(&msg)
+
+	return publish, new(finalize)
+}
+
+func mappingMessage(body []byte, conf *config.Config) ([]byte, interface{}) {
+	var message finalize
+	if err := json.Unmarshal(body, &message); err != nil {
+		return nil, nil
+	}
+	datasetID := uuid.NewSHA1(
+		uuid.NewSHA1(uuid.NameSpaceDNS, []byte(conf.Orchestrator.ProjectFQDN)),
```

Comment:
if we receive the Accession ids in a the `accession_ids` in a different order for the same message, will this generate the same id ?

maybe in a case where we an accession_id did not propagate soon enough and we need to replay the message will that give the same id

Reply:
The response to the first part is the same as above. It is unclear whether reproducibility is an issue here. Of course, I could pick out specific parts of the message to generate the UUID, but in the end, this code will likely call an ID-generating service instead.

I do not understand the second part of the comment, sorry.